src/io/unix/sync.rs

```diff
@@ -0,0 +1,197 @@
+// Sync linux
+
+use crate::prelude::*;
+use bytes::{Bytes, BytesMut};
+use nix::errno::Errno;
+use nix::fcntl::FcntlArg;
+use std::sync::atomic::AtomicU64;
+use std::{
+    os::unix::prelude::{AsRawFd, FileExt},
+    time::SystemTime,
+};
+use tokio::fs::OpenOptions;
+
+/// IO driver for file operations
+#[derive(Debug, Clone, Default)]
+pub struct IODriver;
+
+impl IODriver {
+    pub(crate) async fn open(&self, path: impl AsRef<Path>) -> IOResult<File> {
+        File::from_file(path, |f| f.create(false).append(true).read(true)).await
+    }
+
+    pub(crate) async fn create(&self, path: impl AsRef<Path>) -> IOResult<File> {
+        File::from_file(path, |f| f.create(true).write(true).read(true)).await
+    }
+}
+
+// This size was selected based on results of tests
+// These tests were performed on 8-core machine with HDD
+// Results are confirmed on 8 and 80 threads, for put and get
+const MAX_SYNC_OPERATION_SIZE: usize = 81_920;
+
+#[derive(Debug, Clone)]
+pub(crate) struct File {
+    pub(super) no_lock_fd: Arc<StdFile>,
+    pub(super) size: Arc<AtomicU64>,
+}
+
+#[derive(PartialEq, Eq)]
+enum LockAcquisitionResult {
+    Acquired,
+    AlreadyLocked,
+    Error(Errno),
+}
+
+impl File {
+    pub(crate) fn size(&self) -> u64 {
+        self.size.load(ORD)
+    }
+
+    pub(crate) async fn write_append_all(&self, buf: Bytes) -> IOResult<()> {
+        let offset = self.size.fetch_add(buf.len() as u64, Ordering::SeqCst);
+        self.write_all_at(offset, buf).await
+    }
+
+    pub(crate) async fn write_append_all_buffers(
+        &self,
+        first_buf: Bytes,
+        second_buf: Bytes,
+    ) -> IOResult<()> {
+        let first_len = first_buf.len() as u64;
+        let second_len = second_buf.len() as u64;
+        let len = first_len + second_len;
+        let mut offset = self.size.fetch_add(len, Ordering::SeqCst);
+        let file = self.no_lock_fd.clone();
+        if len <= MAX_SYNC_OPERATION_SIZE as u64 {
+            Self::inplace_sync_call(move || {
+                file.write_all_at(&first_buf, offset)?;
+                offset = offset + first_len;
+                file.write_all_at(&second_buf, offset)?;
+                Ok(())
+            })
+        } else {
+            Self::background_sync_call(move || {
+                file.write_all_at(&first_buf, offset)?;
+                offset = offset + first_len;
+                file.write_all_at(&second_buf, offset)?;
+                Ok(())
+            })
+            .await
+        }
+    }
+
+    pub(crate) async fn write_all_at(&self, offset: u64, buf: Bytes) -> IOResult<()> {
```

Comment:
Please, add `debug_assert!(offset + buf.len() <= self.size());`. We should be sure that this function will not change file size